### DISCUSSION TITLE
langchain with LLM

### DISCUSSION THREAD

**Discussion by woo320** [community/community#82413] [Programming Help]
### Body

from langchain.document_loaders import PyPDFLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.embeddings import HuggingFaceEmbeddings
from langchain.vectorstores import Chroma
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline
from langchain import HuggingFacePipeline
from langchain.prompts import PromptTemplate
from langchain.retrievers.document_compressors import EmbeddingsFilter
from langchain.retrievers import ContextualCompressionRetriever
from langchain.chains import LLMChain

loader = PyPDFLoader('재난법2.pdf')
data = loader.load()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=0)
texts = text_splitter.split_documents(data)

embeddings = HuggingFaceEmbeddings(
model_name="jhgan/ko-sroberta-multitask", encode_kwargs={'normalize_embeddings': True}
)

docsearch = Chroma.from_documents(texts, embeddings)

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-2-7b-chat-hf")
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
llm = HuggingFacePipeline(
pipeline=pipeline,
model_kwargs={"temperature": 0, "max_length": 512},
)

template = """Question: {question}

Answer: Let's work this out in a step by step way to be sure we have the right answer."""
prompt = PromptTemplate(template=template, input_variables=["question"])

llm_chain = LLMChain(llm=llm, prompt=prompt)
question = "자연재난이 뭐야?"
answer = llm_chain.run(
input_document=texts,
question=question,
)
print(answer)

TypeError Traceback (most recent call last)
in <cell line: 13>()
11 llm_chain = LLMChain(llm=llm, prompt=prompt)
12 question = "자연재난이 뭐야?"
---> 13 answer = llm_chain.run(
14 input_document=texts,
15 question=question,

10 frames
[/usr/local/lib/python3.10/dist-packages/transformers/pipelines/init.py](https://localhost:8080/#) in pipeline(task, model, config, tokenizer, feature_extractor, image_processor, framework, revision, use_fast, token, device, device_map, torch_dtype, trust_remote_code, model_kwargs, pipeline_class, **kwargs)
806
807 # Retrieve the task
--> 808 if task in custom_tasks:
809 normalized_task = task
810 targeted_task, task_options = clean_custom_task(custom_tasks[task])

TypeError: unhashable type: 'list'

Please help.

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by faridnec** [comment]
the `task` parameter, which you've set as a list in the pipeline function, might be causing the error. The error suggests that a list is being used in a context where a hashable type is expected.
in the line
```python
pipe = pipeline("text-generation", model=model, tokenizer=tokenizer)
```
Make sure that you're passing a single string as the first argument for the pipeline function. If you intend to perform multiple tasks, you can try to create separate pipelines for each task.
```python
pipe_text_generation = pipeline("text-generation", model=model, tokenizer=tokenizer)
pipe_summarization = pipeline("summarization", model=model, tokenizer=tokenizer)
# ..........................
```
also, make sure that the `llm_chain` and other components of your code are configured to work with a single task or a specific pipeline.
